README.md
# @soroooosh//badi_date

A library to convert Date to Badi date and back.

## Install

```
$ npm install @soroooosh/badi_date
```

## Usage Example

```js
const date1 = BadiDate(day: 1, month: 1, year: 178);
const date2 = BadiDate.fromDate(new Date());

const start = date1.startDateTime;
const end = date1.endDateTime;
const holyDay = date1.holyDay;
const isPeriodOfFast = date2.isPeriodOfFast;
const ayyamIHa = date2.isAyyamIHa;
const nextHolyDay = date2.nextHolyDate;

```


## Known Issues
Tests depends on the local time zone and break in most time zones


CHANGELOG.md
## 1.0.2
* Fix issue with sunset date in the west

## 1.0.1
* Fix readme.md

## 1.0.0

* Initial Open Source release.


src/__test__/badi_date.test.ts
import { BadiDate } from '../lib/badi_date';
import { BahaiHolyDayEnum, bahaiHolyDays } from '../lib/bahai_holyday';

describe('throws errors', () => {
    it('Throws for invalid day and month', () => {
        expect(() => new BadiDate(1, -1, 1, false, null, null)).toThrow('-1 month - Month must be in the range [0-19]');
        expect(() => new BadiDate(1, 20, 1, false, null, null)).toThrow('20 month - Month must be in the range [0-19]');
        expect(() => new BadiDate(1, 1, 0, false, null, null)).toThrow('0 day - Day must be in the range [1-19]');
        expect(() => new BadiDate(1, 1, 20, false, null, null)).toThrow('20 day - Day must be in the range [1-19]');
        expect(() => new BadiDate(1, 1, 4, true, null, null)).toThrow('1 month - Please set month to 0 or set AyyamIHa false');
    });

    it('Throws unsupported years', () => {
        expect(() =>
            new BadiDate(222, 1, 1, false, null, null)).toThrow('Years greater than 221 are not supported yet');
        expect(() => BadiDate.fromDate(new Date(2065, 2, 22), null, null)).toThrow('Dates after 2064-03-19 are not supported yet.');
    });
});

describe('date today', () => {
    const badiDate = new BadiDate(177, 16, 6, false, null, null);
    it('year, month, day', () => {
        const expected = new Date(2021, 0, 4);
        expect(badiDate.getEndDate().getFullYear()).toBe(expected.getFullYear());
        expect(badiDate.getEndDate().getMonth()).toBe(expected.getMonth());
        expect(badiDate.getEndDate().getDate()).toBe(expected.getDate());
    });

    it('after sunset', () => {
        const expected = new Date(2021, 0, 3);
        expect(badiDate.getStartDate().getFullYear()).toBe(expected.getFullYear());
        expect(badiDate.getStartDate().getMonth()).toBe(expected.getMonth());
        expect(badiDate.getStartDate().getDate()).toBe(expected.getDate());
    });

    it('year in Vahid, Vahid, Kull-i-shay', () => {
        expect(badiDate.getKullIShay()).toBe(1);
        expect(badiDate.getVahid()).toBe(10);
        expect(badiDate.getYearInVahid()).toBe(6);
    });

    it('feast, fast, Ayyam-i-ha, Holy day', () => {
        expect(badiDate.isAyyamIHa()).toBe(false);
        expect(badiDate.isFeastDay()).toBe(false);
        expect(badiDate.isPeriodOfFast()).toBe(false);
        expect(badiDate.getHolyDay()).toBe(BahaiHolyDayEnum.NONE);
    });

    it('next feast, next holy day, day of the year', () => {
        expect(
            badiDate.getNextHolyDate()).toEqual(new BadiDate(178, 1, 1, false, null, null));
        expect(badiDate.getNextFeast()).toEqual(new BadiDate(177, 17, 1, false, null, null));
        expect(badiDate.getDayOfYear()).toBe(291);
    });

    it('last AyyamIHa day', () => {
        expect(badiDate.getLastAyyamIHaDayOfYear().getEndDate())
            .toEqual(new Date(2021, 1, 28, 18));
        expect(badiDate.getLastAyyamIHaDayOfYear().getStartDate())
            .toEqual(new Date(2021, 1, 27, 18));
    });
});

describe('naw ruz', () => {
    for (let i = 170; i < 179; i++) {
        it('naw ruz in gregorian date for the year ' + i, () => {
            const date = new BadiDate(i, 1, 1, false, null, null);
            expect(date.getHolyDay()).toBe(BahaiHolyDayEnum.NAW_RUZ);
            expect(date.getStartDate().getMonth()).toBe(2);
            expect(date.getStartDate().getFullYear()).toBe(i + 1843);
            if (i < 173 || [175, 176].includes(i)) {
                expect(date.getEndDate().getDate()).toBe(21);
                expect(BadiDate.getDayOfNawRuz(i)).toBe(21);
            } else {
                expect(date.getEndDate().getDate()).toBe(20);
                expect(BadiDate.getDayOfNawRuz(i)).toBe(20);
            }
        });
    }
});

describe('feasts and holy days', () => {
    const badiDate = new BadiDate(177, 19, 1, false, null, null);
    it('calculates the feasts', () => {
        expect(badiDate.getEndDate()).toEqual(new Date(2021, 2, 1, 18));
        let date = badiDate.getNextFeast();
        expect(date.getEndDate()).toEqual(new Date(2021, 2, 20, 18));
        date = date.getNextFeast();
        expect(date.getEndDate()).toEqual(new Date(2021, 3, 8, 18));
        date = date.getNextFeast();
        expect(date.getEndDate()).toEqual(new Date(2021, 3, 27, 18));
        date = date.getNextFeast();
        expect(date.getEndDate()).toEqual(new Date(2021, 4, 16, 18));
        date = date.getNextFeast();
        expect(date.getEndDate()).toEqual(new Date(2021, 5, 4, 18));
        date = date.getNextFeast();
        expect(date.getEndDate()).toEqual(new Date(2021, 5, 23, 18));
        date = date.getNextFeast();
        expect(date.getEndDate()).toEqual(new Date(2021, 6, 12, 18));
        date = date.getNextFeast();
        expect(date.getEndDate()).toEqual(new Date(2021, 6, 31, 18));
        date = date.getNextFeast();
        expect(date.getEndDate()).toEqual(new Date(2021, 7, 19, 18));
        date = date.getNextFeast();
        expect(date.getEndDate()).toEqual(new Date(2021, 8, 7, 18));
        date = date.getNextFeast();
        expect(date.getEndDate()).toEqual(new Date(2021, 8, 26, 18));
        date = date.getNextFeast();
        expect(date.getEndDate()).toEqual(new Date(2021, 9, 15, 18));
        date = date.getNextFeast();
        expect(date.getEndDate()).toEqual(new Date(2021, 10, 3, 18));
        date = date.getNextFeast();
        expect(date.getEndDate()).toEqual(new Date(2021, 10, 22, 18));
        date = date.getNextFeast();
        expect(date.getEndDate()).toEqual(new Date(2021, 11, 11, 18));
        date = date.getNextFeast();
        expect(date.getEndDate()).toEqual(new Date(2021, 11, 30, 18));
        date = date.getNextFeast();
        expect(date.getEndDate()).toEqual(new Date(2022, 0, 18, 18));
        date = date.getNextFeast();
        expect(date.getEndDate()).toEqual(new Date(2022, 1, 6, 18));
        date = date.getNextFeast();
        expect(date.getEndDate()).toEqual(new Date(2022, 2, 2, 18));
        date = date.getNextFeast();
        expect(date.getEndDate()).toEqual(new Date(2022, 2, 21, 18));
    });

    it('calculates the holy days', () => {
        let date = badiDate.getNextHolyDate();
        expect(date.getEndDate()).toEqual(new Date(2021, 2, 20, 18));
        expect(date.getHolyDay()).toBe(BahaiHolyDayEnum.NAW_RUZ);

        date = date.getNextHolyDate();
        expect(date.getEndDate()).toEqual(new Date(2021, 3, 20, 18));
        expect(date.getHolyDay()).toBe(BahaiHolyDayEnum.RIDVAN1ST);

        date = date.getNextHolyDate();
        expect(date.getEndDate()).toEqual(new Date(2021, 3, 28, 18));
        expect(date.getHolyDay()).toBe(BahaiHolyDayEnum.RIDVAN9TH);

        date = date.getNextHolyDate();
        expect(date.getEndDate()).toEqual(new Date(2021, 4, 1, 18));
        expect(date.getHolyDay()).toBe(BahaiHolyDayEnum.RIDVAN12TH);

        date = date.getNextHolyDate();
        expect(date.getEndDate()).toEqual(new Date(2021, 4, 23, 18));
        expect(date.getHolyDay()).toBe(BahaiHolyDayEnum.DECLARATION_OF_THE_BAB);

        date = date.getNextHolyDate();
        expect(date.getEndDate()).toEqual(new Date(2021, 4, 28, 18));
        expect(date.getHolyDay()).toBe(BahaiHolyDayEnum.ASCENSION_OF_BAHAULLAH);

        date = date.getNextHolyDate();
        expect(date.getEndDate()).toEqual(new Date(2021, 6, 9, 18));
        expect(date.getHolyDay()).toBe(BahaiHolyDayEnum.MARTYRDOM_OF_THE_BAB);

        date = date.getNextHolyDate();
        expect(date.getEndDate()).toEqual(new Date(2021, 10, 6, 18));
        expect(date.getHolyDay()).toBe(BahaiHolyDayEnum.BIRTH_OF_THE_BAB);

        date = date.getNextHolyDate();
        expect(date.getEndDate()).toEqual(new Date(2021, 10, 7, 18));
        expect(date.getHolyDay()).toBe(BahaiHolyDayEnum.BIRTH_OF_BAHAULLAH);

        date = date.getNextHolyDate();
        expect(date.getEndDate()).toEqual(new Date(2021, 10, 25, 18));
        expect(date.getHolyDay()).toBe(BahaiHolyDayEnum.DAY_OF_THE_COVENANT);

        date = date.getNextHolyDate();
        expect(date.getEndDate()).toEqual(new Date(2021, 10, 27, 18));
        expect(date.getHolyDay()).toBe(BahaiHolyDayEnum.ASCENSION_OF_ABDUL_BAHA);
    });
});

describe('corner cases', () => {
    it('round trip naw ruz', () => {
        let date = new Date(2021, 2, 19, 20);
        let badiDate = BadiDate.fromDate(date, null, null);
        expect(badiDate.getStartDate()).toEqual(new Date(2021, 2, 19, 18));
        expect(badiDate.getEndDate()).toEqual(new Date(2021, 2, 20, 18));
        expect(badiDate.month).toBe(1);
        expect(badiDate.day).toBe(1);
        expect(badiDate.year).toBe(178);

        date = new Date(2021, 2, 19, 16);
        badiDate = BadiDate.fromDate(date, null, null);
        expect(badiDate.getStartDate()).toEqual(new Date(2021, 2, 18, 18));
        expect(badiDate.getEndDate()).toEqual(new Date(2021, 2, 19, 18));
        expect(badiDate.month).toBe(19);
        expect(badiDate.day).toBe(19);
        expect(badiDate.year).toBe(177);

        date = new Date(2021, 2, 20, 16);
        badiDate = BadiDate.fromDate(date, null, null);
        expect(badiDate.getStartDate()).toEqual(new Date(2021, 2, 19, 18));
        expect(badiDate.getEndDate()).toEqual(new Date(2021, 2, 20, 18));
        expect(badiDate.month).toBe(1);
        expect(badiDate.day).toBe(1);
        expect(badiDate.year).toBe(178);
    });

    it('new years eve', () => {
        let date = new Date(2020, 11, 31, 16);
        let badiDate = BadiDate.fromDate(date, null, null);
        expect(badiDate.getStartDate()).toEqual(new Date(2020, 11, 30, 18));
        expect(badiDate.getEndDate()).toEqual(new Date(2020, 11, 31, 18));
        expect(badiDate.day).toBe(2);
        expect(badiDate.month).toBe(16);
        expect(badiDate.year).toBe(177);

        date = new Date(2020, 11, 31, 20);
        badiDate = BadiDate.fromDate(date, null, null);
        expect(badiDate.getStartDate()).toEqual(new Date(2020, 11, 31, 18));
        expect(badiDate.getEndDate()).toEqual(new Date(2021, 0, 1, 18));
        expect(badiDate.day).toBe(3);
        expect(badiDate.month).toBe(16);
        expect(badiDate.year).toBe(177);

        date = new Date(2021, 0, 1, 16);
        badiDate = BadiDate.fromDate(date, null, null);
        expect(badiDate.getStartDate()).toEqual(new Date(2020, 11, 31, 18));
        expect(badiDate.getEndDate()).toEqual(new Date(2021, 0, 1, 18));
        expect(badiDate.day).toBe(3);
        expect(badiDate.month).toBe(16);
        expect(badiDate.year).toBe(177);
    });

    it('leap day', () => {
        let date = new Date(2020, 1, 29);
        let badiDate = BadiDate.fromDate(date, null, null);
        expect(badiDate.year).toBe(176);
        expect(badiDate.month).toBe(0);
        expect(badiDate.isAyyamIHa()).toBe(true);
        expect(badiDate.isPeriodOfFast()).toBe(false);
        expect(badiDate.day).toBe(4);

        date = new Date(2020, 2, 1);
        badiDate = BadiDate.fromDate(date, null, null);
        expect(badiDate.isAyyamIHa()).toBe(false);
        expect(badiDate.isPeriodOfFast()).toBe(true);
        expect(badiDate.day).toBe(1);

        date = new Date(2021, 1, 28);
        badiDate = BadiDate.fromDate(date, null, null);
        expect(badiDate.isAyyamIHa()).toBe(true);
        expect(badiDate.isPeriodOfFast()).toBe(false);
        expect(badiDate.day).toBe(4);

        date = new Date(2021, 2, 1);
        badiDate = BadiDate.fromDate(date, null, null);
        expect(badiDate.isAyyamIHa()).toBe(false);
        expect(badiDate.isPeriodOfFast()).toBe(true);
        expect(badiDate.day).toBe(1);

        
        date = new Date(2022, 2, 1);
        badiDate = BadiDate.fromDate(date, null, null);
        expect(badiDate.isAyyamIHa()).toBe(true);
        expect(badiDate.isPeriodOfFast()).toBe(false);
        expect(badiDate.day).toBe(5);

        date = new Date(2022, 2, 2);
        badiDate = BadiDate.fromDate(date, null, null);
        expect(badiDate.isAyyamIHa()).toBe(false);
        expect(badiDate.isPeriodOfFast()).toBe(true);
        expect(badiDate.day).toBe(1);

        
        date = new Date(2004, 2, 1);
        badiDate = BadiDate.fromDate(date, null, null);
        expect(badiDate.isAyyamIHa()).toBe(true);
        expect(badiDate.isPeriodOfFast()).toBe(false);
        expect(badiDate.day).toBe(5);
    });
});

describe("sunset calculation", () => {
    it("handles pols and nonsense values", () => {
        const expected = new Date(2021, 0, 17, 18);
        expect(new BadiDate(177, 17, 1, false, null, null).getStartDate()).
            toEqual(expected);
        expect(
            new BadiDate(177, 17, 1, false, 53.6, null).getStartDate())
            .toEqual(expected);
        expect(
            new BadiDate(177, 17, 1, false, null, 10.0).getStartDate())
            .toEqual(expected);
        expect(
            new BadiDate(
                177,
                17,
                1,
                false,
                53.6,
                190.0,
            )
                .getStartDate())
            .toEqual(expected);
        expect(
            new BadiDate(
                177,
                17,
                1,
                false,
                53.6,
                -190.0)
                .getStartDate())
            .toEqual(expected);
        expect(
            new BadiDate(
                177, 17, 1, false, 66.6, 10.0)
                .getStartDate())
            .toEqual(expected);
        expect(
            new BadiDate(
                177,
                17,
                1,
                false,
                -66.6,
                10.0
            )
                .getStartDate())
            .toEqual(expected);
    });

    // This test only works if your local time is CET or CEST
    it("calculates sunset", () => {
        const expected = new Date(Date.UTC(2021, 0, 17, 15, 34, 40, 947));
        expect(
            new BadiDate(177, 17, 1, false, 53.6, 10.0)
                .getStartDate())
            .toEqual(expected);
    });

    // This test only works if your local time is CET or CEST
    it("handles daylight saving", () => {
        const expected = new Date(Date.UTC(2021, 5, 23, 19, 55));
        const received = new BadiDate(178, 6, 1, false, 53.6, 10.0)
            .getEndDate();
        expect(received.getFullYear()).toBe(expected.getFullYear());
        expect(received.getMonth()).toBe(expected.getMonth());
        expect(received.getUTCDate()).toBe(expected.getUTCDate());
        expect(received.getUTCHours()).toBe(expected.getUTCHours());
        expect(received.getUTCMinutes()).toBe(expected.getUTCMinutes());
    });

    it("handles sunset on next Gregorian day", () => {
        const expected = new Date(Date.UTC(2021, 5, 23, 22, 18));
        const received = new BadiDate(178, 6, 1, false, 64.6, 8.0)
            .getEndDate();
        expect(received.getFullYear()).toBe(expected.getFullYear());
        expect(received.getMonth()).toBe(expected.getMonth());
        expect(received.getUTCDate()).toBe(expected.getUTCDate());
        expect(received.getUTCHours()).toBe(expected.getUTCHours());
        expect(received.getUTCMinutes()).toBe(expected.getUTCMinutes());
    });
});


src/lib/years.ts
class YearSpecifics {
  readonly year: number;
  readonly leapday: boolean;
  readonly nawRuzOnMarch21: boolean;
  readonly birthOfBab: number;

  constructor(year: number,
    birthOfBab: number,
    leapday: boolean,
    nawRuzOnMarch21: boolean) {
    this.year = year;
    this.leapday = leapday;
    this.birthOfBab = birthOfBab;
    this.nawRuzOnMarch21 = nawRuzOnMarch21;
  };

  static createNawRuzOn21Year(year: number,
    birthOfBab: number): YearSpecifics {
    return new YearSpecifics(year, birthOfBab, false, true);
  }

  static createYearNotLeap(year: number,
    birthOfBab: number): YearSpecifics {
    return new YearSpecifics(year, birthOfBab, false, false);
  }

  static createLeapYear(year: number,
    birthOfBab: number): YearSpecifics {
    return new YearSpecifics(year, birthOfBab, true, false);
  }
}

export const yearSpecifics = new Map<number, YearSpecifics>([
  [172, YearSpecifics.createNawRuzOn21Year(172, 238)],
  [173, YearSpecifics.createYearNotLeap(173, 227)],
  [174, YearSpecifics.createLeapYear(174, 216)],
  [175, YearSpecifics.createNawRuzOn21Year(175, 234)],
  [176, YearSpecifics.createNawRuzOn21Year(176, 223)],
  [177, YearSpecifics.createYearNotLeap(177, 213)],
  [178, YearSpecifics.createLeapYear(178, 232)],
  [179, YearSpecifics.createNawRuzOn21Year(179, 220)],
  [180, YearSpecifics.createNawRuzOn21Year(180, 210)],
  [181, YearSpecifics.createYearNotLeap(181, 228)],
  [182, YearSpecifics.createLeapYear(182, 217)],
  [183, YearSpecifics.createNawRuzOn21Year(183, 235)],
  [184, YearSpecifics.createNawRuzOn21Year(184, 224)],
  [185, YearSpecifics.createYearNotLeap(185, 214)],
  [186, YearSpecifics.createYearNotLeap(186, 233)],
  [187, YearSpecifics.createLeapYear(187, 223)],
  [188, YearSpecifics.createNawRuzOn21Year(188, 211)],
  [189, YearSpecifics.createYearNotLeap(189, 230)],
  [190, YearSpecifics.createYearNotLeap(190, 238)],
  [191, YearSpecifics.createLeapYear(191, 238)],
  [192, YearSpecifics.createNawRuzOn21Year(192, 226)],
  [193, YearSpecifics.createYearNotLeap(193, 215)],
  [194, YearSpecifics.createYearNotLeap(194, 234)],
  [195, YearSpecifics.createLeapYear(195, 224)],
  [196, YearSpecifics.createNawRuzOn21Year(196, 213)],
  [197, YearSpecifics.createYearNotLeap(197, 232)],
  [198, YearSpecifics.createYearNotLeap(198, 221)],
  [199, YearSpecifics.createLeapYear(199, 210)],
  [200, YearSpecifics.createNawRuzOn21Year(200, 228)],
  [201, YearSpecifics.createYearNotLeap(201, 217)],
  [202, YearSpecifics.createYearNotLeap(202, 236)],
  [203, YearSpecifics.createLeapYear(203, 225)],
  [204, YearSpecifics.createNawRuzOn21Year(204, 214)],
  [205, YearSpecifics.createYearNotLeap(205, 233)],
  [206, YearSpecifics.createYearNotLeap(206, 223)],
  [207, YearSpecifics.createLeapYear(207, 212)],
  [208, YearSpecifics.createNawRuzOn21Year(208, 230)],
  [209, YearSpecifics.createYearNotLeap(209, 219)],
  [210, YearSpecifics.createYearNotLeap(210, 237)],
  [211, YearSpecifics.createLeapYear(211, 227)],
  [212, YearSpecifics.createNawRuzOn21Year(212, 215)],
  [213, YearSpecifics.createYearNotLeap(213, 234)],
  [214, YearSpecifics.createYearNotLeap(214, 224)],
  [215, YearSpecifics.createYearNotLeap(215, 213)],
  [216, YearSpecifics.createLeapYear(216, 232)],
  [217, YearSpecifics.createYearNotLeap(217, 220)],
  [218, YearSpecifics.createYearNotLeap(218, 209)],
  [219, YearSpecifics.createYearNotLeap(219, 228)],
  [220, YearSpecifics.createLeapYear(220, 218)],
  [221, YearSpecifics.createYearNotLeap(221, 236)],
]);


src/lib/bahai_holyday.ts
export enum BahaiHolyDayEnum {
  // Naw-Ruz
  NAW_RUZ,

  // 1st day of Ridvan
  RIDVAN1ST,

  // 9th day of Ridvan
  RIDVAN9TH,

  // 12th day of Ridvan
  RIDVAN12TH,

  // Declaration of the Bab
  DECLARATION_OF_THE_BAB,

  // Ascension of Baha'u'llah
  ASCENSION_OF_BAHAULLAH,

  // Martyrdom of the Bab
  MARTYRDOM_OF_THE_BAB,

  // Birth of the Bab
  BIRTH_OF_THE_BAB,

  // Birth of Baha'u'llah
  BIRTH_OF_BAHAULLAH,

  // Day of the Covenant (not a holy day)
  DAY_OF_THE_COVENANT,

  // Ascension of Abdu'l-Baha (not a holy day)
  ASCENSION_OF_ABDUL_BAHA,

  // Not a Holyday
  NONE,
}

class BahaiHolyDay {
  readonly dayOfYear: number;
  readonly type: BahaiHolyDayEnum;

  constructor(dayOfYear: number, type: BahaiHolyDayEnum) {
    this.dayOfYear = dayOfYear;
    this.type = type;
  };

  getDayOfTheYear(dayOfYearBirthOfBab: number): number {
    if (this.type == BahaiHolyDayEnum.BIRTH_OF_THE_BAB) {
      return dayOfYearBirthOfBab ?? this.dayOfYear;
    } else if (this.type == BahaiHolyDayEnum.BIRTH_OF_BAHAULLAH &&
      dayOfYearBirthOfBab) {
      return dayOfYearBirthOfBab + 1;
    }
    return this.dayOfYear;
  }
}

export const bahaiHolyDays: BahaiHolyDay[] = [
  // Naw-Ruz
  new BahaiHolyDay(1, BahaiHolyDayEnum.NAW_RUZ),

  // 1st day of Ridvan
  new BahaiHolyDay(32, BahaiHolyDayEnum.RIDVAN1ST),

  // 9th day of Ridvan
  new BahaiHolyDay(40, BahaiHolyDayEnum.RIDVAN9TH),

  // 12th day of Ridvan
  new BahaiHolyDay(43, BahaiHolyDayEnum.RIDVAN12TH),

  // Decleration of the Bab
  new BahaiHolyDay(65, BahaiHolyDayEnum.DECLARATION_OF_THE_BAB),

  // Ascension of Baha'u'llah
  new BahaiHolyDay(70, BahaiHolyDayEnum.ASCENSION_OF_BAHAULLAH),

  // Martyrdom of the Bab
  new BahaiHolyDay(112, BahaiHolyDayEnum.MARTYRDOM_OF_THE_BAB),

  // Birth of the Bab
  new BahaiHolyDay(214, BahaiHolyDayEnum.BIRTH_OF_THE_BAB),

  // Birth of Baha'u'llah
  new BahaiHolyDay(237, BahaiHolyDayEnum.BIRTH_OF_BAHAULLAH),

  // Day of the Covenant (not a holy day)
  new BahaiHolyDay(251, BahaiHolyDayEnum.DAY_OF_THE_COVENANT),

  // Ascension of Abdu'l-Baha (not a holy day)
  new BahaiHolyDay(253, BahaiHolyDayEnum.ASCENSION_OF_ABDUL_BAHA),
];


src/lib/badi_date.ts
import SunCalc = require('suncalc');
import { BahaiHolyDayEnum, bahaiHolyDays } from './bahai_holyday';
import { yearSpecifics } from './years';


type nullableNumber = number | null;

const LAST_YEAR_SUPPORTED = 221;
const DAY_IN_MILLISECONDS = 86400000;
const YEAR_ONE_IN_GREGORIAN = 1844;
const YEAR_ZERO_IN_GREGORIAN = YEAR_ONE_IN_GREGORIAN - 1;

// A Badi Date
export class BadiDate {
  readonly day: number;

  // The month number with Baha = 1, ... Ayyam'i'Ha = 19, and Ala = 20
  _monthIntern = -1;

  // The month number with Ayyam'i'Ha = 0 and Baha = 1, ... Ala = 19
  readonly month: number;

  // The full year of the Badi Calendar with 2020 AC = 177 BE
  readonly year: number;

  // longitude value of degree coordinates for sunset calculation in the range [-180,180]
  readonly longitude: nullableNumber;

  // latitude value of degree coordinates for sunset calculation in the range [-90,90]
  readonly latitude: nullableNumber;

  // day of the badi year
  readonly dayOfYear: number;

  // Badi date
  // for now only for the years up to LAST_YEAR_SUPPORTED
  // Dates before the year 172 are calculated according to the Baha'i Calendar
  // used in western countries.
  // parameters:
  // day int in range [1-19]
  // month int in range [0-19]
  // year int
  // longitude and latitude double for sunset calculation
  // ayyamIHa bool
  // For Ayyam'i'Ha set month to 0 or leave it empty and set ayyamIHa to true
  constructor(
    year: number,
    month: number,
    day: number,
    ayyamIHa: boolean,
    latitude: nullableNumber,
    longitude: nullableNumber) {
    this.year = year;
    this.month = month;
    this.day = day;
    this.latitude = latitude;
    this.longitude = longitude;
    if (day < 1 || day > 19) {
      throw new Error(day + ' day - Day must be in the range [1-19]');
    }
    if (month < 0 || month > 19) {
      throw new Error(month + ' month - Month must be in the range [0-19]');
    }
    if (month !== 0 && ayyamIHa) {
      throw new Error(month + ' month - Please set month to 0 or set AyyamIHa false');
    }
    if (year > LAST_YEAR_SUPPORTED) {
      throw new Error(
        'Years greater than ' + LAST_YEAR_SUPPORTED + ' are not supported yet');
    }
    this._monthIntern = month === 0
      ? 19
      : month === 19
        ? 20
        : month;

    this.dayOfYear = this.getDayOfYear();
  }

  // The year in the Vahid. A value in the range from [1-19]
  getYearInVahid(): number {
    return this.year % 19 === 0 ? 19 : this.year % 19;
  }

  // Vahid = 19 years
  getVahid(): number {
    return Math.floor((this.year / 19)) + 1;
  }

  // Kull'i'shay = 19 Vahids = 361 years
  getKullIShay(): number {
    return Math.floor((this.year / 361)) + 1;
  }

  // Number of Ayyam'i'ha days the year has
  // For years < 172: use only for January 1st to before Naw-Ruz
  static _getNumberAyyamIHaDays(year: number): number {
    const yearSpecific = yearSpecifics.get(year);
    if (yearSpecific == null) {
      const gregYear = year + YEAR_ONE_IN_GREGORIAN;
      const isLeapYear =
        gregYear % 4 === 0 && gregYear % 100 !== 0 || gregYear % 400 === 0;
      return isLeapYear ? 5 : 4;
    }
    return yearSpecific.leapday ? 5 : 4;
  }

  static getDayOfNawRuz(year: number): number {
    const yearSpecific = yearSpecifics.get(year);
    if (yearSpecific == null) {
      return 21;
    }
    return yearSpecific.nawRuzOnMarch21 ? 21 : 20;
  }

  // the day of the year with Naw Ruz = 1
  getDayOfYear(): number {
    if (this._monthIntern === 20) {
      return 342 + BadiDate._getNumberAyyamIHaDays(this.year) + this.day;
    }
    return (this._monthIntern - 1) * 19 + this.day;
  }

  // Is the date in the period of fast
  isPeriodOfFast(): boolean {
    return this.month === 19;
  }

  // is the date an Ayyam'i'Ha day
  isAyyamIHa(): boolean {
    return this.month === 0;
  }

  // is the date a feast date
  isFeastDay(): boolean {
    return this.day === 1 && !this.isAyyamIHa;
  }

  static _calculateSunSet(date: Date,
    longitude: nullableNumber, latitude: nullableNumber): Date {
    const fallback = new Date(date.getFullYear(), date.getMonth(), date.getDate(), 18);
    // return 6pm if no location or if in the poles
    if (latitude === null ||
      longitude === null ||
      latitude > 66.0 ||
      latitude < -66.0 ||
      Math.abs(longitude) > 180.0) {
      return fallback;
    }
    const sunCalcTimes = SunCalc.getTimes(date,
      latitude, longitude);
    // The sunset of places far west might have the sunset calculated
    // for the day before. In that case we add a day and calculate again.
    if (sunCalcTimes.sunset?.getDate() === date.getDate() - 1) {
      const sunCalcWithAdjustment = SunCalc.getTimes(
        new Date(date.getFullYear(), date.getMonth(), date.getDate() + 1),
        latitude,
        longitude);
      return sunCalcWithAdjustment.sunset ?? fallback;
    }
    return sunCalcTimes.sunset ?? fallback;
  }

  getNawRuzDate(): Date {
    return new Date(Date.UTC(this.year + YEAR_ZERO_IN_GREGORIAN, 2, BadiDate.getDayOfNawRuz(this.year)));
  }

  // Start Date
  getStartDate(): Date {
    const date = this.getNawRuzDate();
    date.setUTCDate(date.getUTCDate() + this.dayOfYear - 2);
    return BadiDate._calculateSunSet(date, this.longitude, this.latitude);
  }

  // End Date
  getEndDate(): Date {
    const date = this.getNawRuzDate();
    date.setUTCDate(date.getUTCDate() + this.dayOfYear - 1);
    return BadiDate._calculateSunSet(date, this.longitude, this.latitude);
  }

  static _fromYearAndDayOfYear(
    year: number,
    doy: number,
    longitude: nullableNumber,
    latitude: nullableNumber): BadiDate {
    if (doy < 1 || doy > 366) {
      throw Error(
        doy + ' doy, Day of year must be in the range [1-366]');
    }
    const month = Math.ceil(doy / 19);
    const day = doy - (month - 1) * 19;
    const numberOfAyyamIHaDays = BadiDate._getNumberAyyamIHaDays(year);
    if (month < 19) {
      return new BadiDate(
        year,
        month,
        day,
        false,
        longitude,
        latitude);
    } else if (month === 19 && day <= numberOfAyyamIHaDays) {
      return new BadiDate(
        year,
        0,
        day,
        true,
        longitude,
        latitude);
    }
    const alaDay = doy - 342 - numberOfAyyamIHaDays;
    return new BadiDate(
      year,
      19,
      alaDay,
      false,
      longitude,
      latitude);
  }

  // BadiDate from a DateTime object
  // Optional parameter double longitude, latitude for the sunset time
  static fromDate(dateTime: Date,
    longitude: nullableNumber, latitude: nullableNumber): BadiDate {
    if (dateTime > new Date(2065, 2, 19)) {
      throw Error('Dates after 2064-03-19 are not supported yet.');
    }
    const isAfterSunset = dateTime > BadiDate._calculateSunSet(dateTime,
      longitude, latitude);
    const date = dateTime;
    if (isAfterSunset) date.setDate(dateTime.getDate() + 1);
    const badiYear = date.getFullYear() - YEAR_ZERO_IN_GREGORIAN;
    const isBeforeNawRuz =
      date < new Date(date.getFullYear(), 2, BadiDate.getDayOfNawRuz(badiYear));
    const adaptedBadiYear = isBeforeNawRuz ? badiYear - 1 : badiYear;
    const lastNawRuz = isBeforeNawRuz
      ? new Date(date.getFullYear() - 1, 2, BadiDate.getDayOfNawRuz(adaptedBadiYear))
      : new Date(date.getFullYear(), 2, BadiDate.getDayOfNawRuz(adaptedBadiYear));
    const doy = Math.floor((date.getTime() - lastNawRuz.getTime()) / DAY_IN_MILLISECONDS);
    // +1 because naw ruz has a doy of 1 but a difference of 0
    return BadiDate._fromYearAndDayOfYear(
      adaptedBadiYear,
      doy + 1,
      longitude,
      latitude);
  }

  // If the BadiDate is a Baha'i Holy day the Holy date else BahaiHolyDayEnum.NONE
  getHolyDay(): BahaiHolyDayEnum {
    const birthOfBab = yearSpecifics.get(this.year)?.birthOfBab ?? 214;
    return bahaiHolyDays
      .find(
        (holyDay) =>
          holyDay?.getDayOfTheYear(birthOfBab) ==
          this.dayOfYear)
      ?.type ?? BahaiHolyDayEnum.NONE;
  }

  // The BadiDate of the next feast
  getNextFeast(): BadiDate {
    if (this.month === 19) {
      return new BadiDate(
        this.year + 1,
        1,
        1,
        false,
        this.longitude,
        this.latitude);
    }
    return new BadiDate(
      this.year,
      this.month + 1,
      1,
      false,
      this.longitude,
      this.latitude);
  }

  // The BadiDate of the next Holy day
  getNextHolyDate(): BadiDate {
    const birthOfBab = yearSpecifics.get(this.year)?.birthOfBab ?? 0;
    const doy = bahaiHolyDays
      .find(
        (holyDay) =>
          (holyDay?.getDayOfTheYear(birthOfBab) ??
            0) >
          this.dayOfYear)
      ?.getDayOfTheYear(birthOfBab);
    if (doy == null) {
      return BadiDate._fromYearAndDayOfYear(
        this.year + 1,
        1,
        this.longitude,
        this.latitude);
    }
    return BadiDate._fromYearAndDayOfYear(
      this.year,
      doy,
      this.longitude,
      this.latitude);
  }

  // return the last Ayyam'i'Ha day of that Badi year
  getLastAyyamIHaDayOfYear(): BadiDate {
    const firstAla = new BadiDate(
      this.year,
      19,
      1,
      false,
      this.longitude,
      this.latitude);
    return BadiDate._fromYearAndDayOfYear(
      this.year,
      firstAla.dayOfYear - 1,
      this.longitude,
      this.latitude);
  }
}
